File: st-exerciseviewer/src/main/java/de/saring/exerciseviewer/gui/panels/MainPanelController.java
package de.saring.exerciseviewer.gui.panels;

import java.time.format.DateTimeFormatter;
import java.time.format.FormatStyle;

import de.saring.exerciseviewer.gui.EVDocument;
import javafx.event.ActionEvent;
import javafx.fxml.FXML;
import javafx.scene.control.ChoiceBox;
import javafx.scene.control.Label;

import de.saring.exerciseviewer.data.EVExercise;
import de.saring.exerciseviewer.data.HeartRateLimit;
import de.saring.exerciseviewer.gui.EVContext;
import de.saring.util.unitcalc.FormatUtils;
import javafx.util.StringConverter;

/**
 * Controller (MVC) class of the "Main" panel, which displays all the main exercise data.
 *
 * @author Stefan Saring
 */
public class MainPanelController extends AbstractPanelController {

    private DiagramPanelController diagramPanelController;

    @FXML
    private Label laDeviceValue;
    @FXML
    private Label laTypeValue;
    @FXML
    private Label laDateTimeValue;
    @FXML
    private Label laDurationValue;
    @FXML
    private Label laEnergyValue;

    @FXML
    private Label laHeartrateAvgValue;
    @FXML
    private Label laHeartrateMaxValue;

    @FXML
    private ChoiceBox<HeartRateLimit> cbHeartrateRanges;
    @FXML
    private Label laTimeBelowValue;
    @FXML
    private Label laTimeWithinValue;
    @FXML
    private Label laTimeAboveValue;

    @FXML
    private Label laModeSpeedValue;
    @FXML
    private Label laModeAltitudeValue;
    @FXML
    private Label laModeCadendeValue;
    @FXML
    private Label laModePowerValue;

    @FXML
    private Label laTotalExerciseTimeValue;
    @FXML
    private Label laTotalRidingTimeValue;
    @FXML
    private Label laTotalEnergyValue;
    @FXML
    private Label laOdometerValue;

    /**
     * Standard c'tor for dependency injection.
     *
     * @param context the ExerciseViewer UI context
     * @param document the ExerciseViewer document / model
     */
    public MainPanelController(final EVContext context, final EVDocument document) {
        super(context, document);
    }

    public void setDiagramPanelController(final DiagramPanelController diagramPanelController) {
        this.diagramPanelController = diagramPanelController;
    }

    @Override
    protected String getFxmlFilename() {
        return "/fxml/panels/MainPanel.fxml";
    }

    @Override
    protected void setupPanel() {
        showExerciseData();
        setupHeartrateRangeSelection();
    }

    private void showExerciseData() {
        final EVExercise exercise = getDocument().getExercise();
        final FormatUtils formatUtils = getContext().getFormatUtils();

        if (exercise.getFileType() != EVExercise.ExerciseFileType.HRM) {

            // fill device and type data
            if (exercise.getDeviceName() != null && !exercise.getDeviceName().trim().isEmpty()) {
                laDeviceValue.setText(exercise.getDeviceName());
            }
            if (exercise.getType() != null) {
                laTypeValue.setText(exercise.getType());
            }

            // fill energy data
            if (exercise.getEnergy() > 0) {
                laEnergyValue.setText(formatUtils.caloriesToString(exercise.getEnergy()));
            }

            // fill statistics data
            if (exercise.getSumExerciseTime() > 0) {
                laTotalExerciseTimeValue.setText(formatUtils.minutes2TimeString(exercise.getSumExerciseTime()));
            }
            if (exercise.getEnergyTotal() > 0) {
                laTotalEnergyValue.setText(formatUtils.caloriesToString(exercise.getEnergyTotal()));
            }

            // fill total riding time (available only in S710 exercises)
            if (exercise.getFileType() == EVExercise.ExerciseFileType.S710RAW) {
                laTotalRidingTimeValue.setText(formatUtils.minutes2TimeString(exercise.getSumRideTime()));
            }
        }

        // fill time data
        if (exercise.getDateTime() != null) {
            laDateTimeValue.setText(exercise.getDateTime().format(
                    DateTimeFormatter.ofLocalizedDateTime(FormatStyle.MEDIUM)));
        }

        if (exercise.getDuration() > 0) {
            laDurationValue.setText(formatUtils.tenthSeconds2TimeString(exercise.getDuration()));
        }

        // fill heartrate data
        if (exercise.getHeartRateAVG() > 0) {
            laHeartrateAvgValue.setText(formatUtils.heartRateToString(exercise.getHeartRateAVG()));
        }
        if (exercise.getHeartRateMax() > 0) {
            laHeartrateMaxValue.setText(formatUtils.heartRateToString(exercise.getHeartRateMax()));
        }

        // fill recording mode data
        laModeSpeedValue.setText(boolean2EnabledString(exercise.getRecordingMode().isSpeed()));
        laModeAltitudeValue.setText(boolean2EnabledString(exercise.getRecordingMode().isAltitude()));
        laModeCadendeValue.setText(boolean2EnabledString(exercise.getRecordingMode().isCadence()));
        laModePowerValue.setText(boolean2EnabledString(exercise.getRecordingMode().isPower()));

        // fill odometer data (if available, e.g. not on Polar S410 or S610)
        if (exercise.getOdometer() != 0) {
            laOdometerValue.setText(formatUtils.distanceToString(exercise.getOdometer(), 2));
        }
    }

    /**
     * Fills the heartrate range selection choicebox with the ranges stored in the exercise and selects
     * the first one. When no ranges are available, the selection will be disabled.<br/>
     */
    private void setupHeartrateRangeSelection() {

        cbHeartrateRanges.addEventHandler(ActionEvent.ACTION, event -> updateHeartRateRangeTimes());

        cbHeartrateRanges.setConverter(new StringConverter<HeartRateLimit>() {
            @Override
            public String toString(final HeartRateLimit limit) {
                final String unitName = limit.isAbsoluteRange() ? "bpm" : "%";
                return "" + limit.getLowerHeartRate() + " - " + limit.getUpperHeartRate() + " " + unitName;
            }

            @Override
            public HeartRateLimit fromString(String string) {
                throw new UnsupportedOperationException();
            }
        });

        final EVExercise exercise = getDocument().getExercise();
        if (exercise.getHeartRateLimits() != null && exercise.getHeartRateLimits().length > 0) {
            cbHeartrateRanges.getItems().addAll(exercise.getHeartRateLimits());
            cbHeartrateRanges.getSelectionModel().select(0);
        } else {
            cbHeartrateRanges.setDisable(true);
        }
    }

    /**
     * Updates the heartrate time value labels with the values of the current selected range.
     */
    private void updateHeartRateRangeTimes() {

        final EVExercise exercise = getDocument().getExercise();
        final FormatUtils formatUtils = getContext().getFormatUtils();
        final HeartRateLimit limit = cbHeartrateRanges.getValue();

        // calculate percentages of times below, within and above
        int percentsBelow = 0, percentsWithin = 0, percentsAbove = 0;
        if (exercise.getDuration() > 0) {
            percentsBelow = (int) Math.round(limit.getTimeBelow() / (double) exercise.getDuration() * 10 * 100);
            percentsWithin = (int) Math.round(limit.getTimeWithin() / (double) exercise.getDuration() * 10 * 100);
            percentsAbove = (int) Math.round(limit.getTimeAbove() / (double) exercise.getDuration() * 10 * 100);
        }

        laTimeBelowValue.setText(formatUtils.seconds2TimeString(limit.getTimeBelow()) + "   (" + percentsBelow + " %)");
        laTimeWithinValue.setText(formatUtils.seconds2TimeString(limit.getTimeWithin()) + "   (" + percentsWithin + " %)");
        laTimeAboveValue.setText(formatUtils.seconds2TimeString(limit.getTimeAbove()) + "   (" + percentsAbove + " %)");

        // update heartrate range in diagram
        diagramPanelController.displayDiagramForHeartrateRange(limit);
    }

    private String boolean2EnabledString(final boolean enabled) {
        return getContext().getResources().getString(enabled ? "common.enabled" : "common.disabled");
    }
}

File: st-exerciseviewer/src/main/kotlin/de/saring/exerciseviewer/gui/panels/MainPanelController.kt
package de.saring.exerciseviewer.gui.panels

import de.saring.exerciseviewer.data.HeartRateLimit
import de.saring.exerciseviewer.gui.EVContext
import de.saring.exerciseviewer.gui.EVDocument
import de.saring.util.StringUtils
import de.saring.util.unitcalc.TimeUtils
import javafx.event.ActionEvent
import javafx.fxml.FXML
import javafx.scene.control.ChoiceBox
import javafx.scene.control.Label
import javafx.util.StringConverter
import java.time.format.DateTimeFormatter
import java.time.format.FormatStyle

/**
 * Controller (MVC) class of the "Main" panel, which displays all the main exercise data.
 *
 * @constructor constructor for dependency injection
 * @param context the ExerciseViewer UI context
 * @param document the ExerciseViewer document / model
 *
 * @author Stefan Saring
 */
class MainPanelController(
        context: EVContext,
        document: EVDocument) : AbstractPanelController(context, document) {

    lateinit var diagramPanelController: DiagramPanelController

    @FXML
    private lateinit var laDateTimeValue: Label
    @FXML
    private lateinit var laDurationValue: Label
    @FXML
    private lateinit var laEnergyValue: Label
    @FXML
    private lateinit var laDeviceValue: Label
    @FXML
    private lateinit var laTypeValue: Label

    @FXML
    private lateinit var laHeartrateAvgValue: Label
    @FXML
    private lateinit var laHeartrateMaxValue: Label

    @FXML
    private lateinit var cbHeartrateRanges: ChoiceBox<HeartRateLimit>
    @FXML
    private lateinit var laTimeBelowValue: Label
    @FXML
    private lateinit var laTimeWithinValue: Label
    @FXML
    private lateinit var laTimeAboveValue: Label

    @FXML
    private lateinit var laModeHeartrateValue: Label
    @FXML
    private lateinit var laModeSpeedValue: Label
    @FXML
    private lateinit var laModeAltitudeValue: Label
    @FXML
    private lateinit var laModeCadendeValue: Label
    @FXML
    private lateinit var laModePowerValue: Label

    @FXML
    private lateinit var laTotalExerciseTimeValue: Label
    @FXML
    private lateinit var laTotalRidingTimeValue: Label
    @FXML
    private lateinit var laTotalEnergyValue: Label
    @FXML
    private lateinit var laOdometerValue: Label


    override val fxmlFilename: String = "/fxml/panels/MainPanel.fxml"

    override fun setupPanel() {
        showExerciseData()
        setupHeartrateRangeSelection()
    }

    private fun showExerciseData() {
        val exercise = document.exercise
        val formatUtils = context.formatUtils

        // fill energy data
        exercise.energy?.let {
            laEnergyValue.text = formatUtils.caloriesToString(it)
        }

        // fill statistics data
        exercise.sumExerciseTime?.let {
            laTotalExerciseTimeValue.text = TimeUtils.minutes2TimeString(it)
        }
        // fill total riding time (available only in S710 exercises)
        exercise.sumRideTime?.let {
            laTotalRidingTimeValue.text = TimeUtils.minutes2TimeString(it)
        }
        exercise.energyTotal?.let {
            laTotalEnergyValue.text = formatUtils.caloriesToString(it)
        }
        // fill odometer data (if available, e.g. not on Polar S410 or S610)
        exercise.odometer?.let {
            laOdometerValue.text = formatUtils.distanceToString(it.toDouble(), 2)
        }

        // fill device and type data
        if (!StringUtils.isNullOrEmpty(exercise.deviceName)) {
            laDeviceValue.text = exercise.deviceName
        }
        exercise.type?.let {
            laTypeValue.text = it
        }

        // fill time data
        exercise.dateTime?.let {
            laDateTimeValue.text = it.format(DateTimeFormatter.ofLocalizedDateTime(FormatStyle.MEDIUM))
        }

        exercise.duration?.let {
            laDurationValue.text = TimeUtils.tenthSeconds2TimeString(it)
        }

        // fill heartrate data
        exercise.heartRateAVG?.let {
            laHeartrateAvgValue.text = formatUtils.heartRateToString(it.toInt())
        }
        exercise.heartRateMax?.let {
            laHeartrateMaxValue.text = formatUtils.heartRateToString(it.toInt())
        }

        // fill recording mode data
        laModeHeartrateValue.text = boolean2EnabledString(exercise.recordingMode.isHeartRate)
        laModeSpeedValue.text = boolean2EnabledString(exercise.recordingMode.isSpeed)
        laModeAltitudeValue.text = boolean2EnabledString(exercise.recordingMode.isAltitude)
        laModeCadendeValue.text = boolean2EnabledString(exercise.recordingMode.isCadence)
        laModePowerValue.text = boolean2EnabledString(exercise.recordingMode.isPower)
    }

    /**
     * Fills the heartrate range selection choicebox with the ranges stored in the exercise and selects
     * the first one. When no ranges are available, the selection will be disabled.
     */
    private fun setupHeartrateRangeSelection() {

        cbHeartrateRanges.addEventHandler(ActionEvent.ACTION) { _ -> updateHeartRateRangeTimes() }

        cbHeartrateRanges.converter = object : StringConverter<HeartRateLimit>() {
            override fun toString(limit: HeartRateLimit): String {
                val unitName = if (limit.isAbsoluteRange) "bpm" else "%"
                return "${limit.lowerHeartRate} - ${limit.upperHeartRate} $unitName"
            }

            override fun fromString(string: String): HeartRateLimit =
                throw UnsupportedOperationException()
        }

        val exercise = document.exercise
        if (!exercise.heartRateLimits.isEmpty()) {
            cbHeartrateRanges.items.addAll(exercise.heartRateLimits)
            cbHeartrateRanges.selectionModel.select(0)
        } else {
            cbHeartrateRanges.setDisable(true)
        }
    }

    /**
     * Updates the heartrate time value labels with the values of the current selected range.
     */
    private fun updateHeartRateRangeTimes() {

        val limit = cbHeartrateRanges.value

        // calculate percentages of times below, within and above
        document.exercise.duration?.let { duration ->
            displayHeartRateRangeTime(limit.timeBelow, duration, laTimeBelowValue)
            displayHeartRateRangeTime(limit.timeWithin, duration, laTimeWithinValue)
            displayHeartRateRangeTime(limit.timeAbove, duration, laTimeAboveValue)
        }

        // update heartrate range in diagram
        diagramPanelController.displayDiagramForHeartrateRange(limit)
    }

    private fun displayHeartRateRangeTime(time: Int?, duration: Int, label: Label) {

        if (time != null) {
            val percents = Math.round(time / duration.toDouble() * 10.0 * 100.0).toInt()
            label.text = TimeUtils.seconds2TimeString(time) + "   ($percents %)"
        } else {
            label.text = context.resources.getString("common.n_a_")
        }
    }

    private fun boolean2EnabledString(enabled: Boolean) =
        context.resources.getString(if (enabled) "common.enabled" else "common.disabled")
}
